File: DSA/Ayush_Mishra_590014844/Day_39_Question_3.c
#include <stdio.h>
#define SIZE 100
int arr[SIZE], top1 = -1, top2 = SIZE;

void push1(int val) {
    if (top1 + 1 == top2) return;
    arr[++top1] = val;
}

void push2(int val) {
    if (top2 - 1 == top1) return;
    arr[--top2] = val;
}

int pop1() {
    if (top1 == -1) return -1;
    return arr[top1--];
}

int pop2() {
    if (top2 == SIZE) return -1;
    return arr[top2++];
}

void print1() {
    for (int i = 0; i <= top1; ++i) printf("%d ", arr[i]);
    printf("\n");
}

void print2() {
    for (int i = SIZE - 1; i >= top2; --i) printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    push1(1);
    push1(2);
    push2(9);
    push2(8);
    print1();
    print2();
    push1(3);
    pop2();
    print1();
    print2();
    return 0;
}

File: DSA/Ayush_Mishra_590014844/Day_39_Question_3.py
class TwoStacks:
    def __init__(self, size):
        self.size = size
        self.arr = [None] * size
        self.top1 = -1
        self.top2 = size

    def push1(self, val):
        if self.top1 + 1 == self.top2:
            return
        self.top1 += 1
        self.arr[self.top1] = val

    def push2(self, val):
        if self.top2 - 1 == self.top1:
            return
        self.top2 -= 1
        self.arr[self.top2] = val

    def pop1(self):
        if self.top1 == -1:
            return None
        val = self.arr[self.top1]
        self.top1 -= 1
        return val

    def pop2(self):
        if self.top2 == self.size:
            return None
        val = self.arr[self.top2]
        self.top2 += 1
        return val

    def print1(self):
        print(*self.arr[:self.top1 + 1])

    def print2(self):
        print(*self.arr[self.top2:][::-1])

s = TwoStacks(100)
s.push1(1)
s.push1(2)
s.push2(9)
s.push2(8)
s.print1()
s.print2()
s.push1(3)
s.pop2()
s.print1()
s.print2()
